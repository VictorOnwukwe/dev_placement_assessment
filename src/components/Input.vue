<template>
  <div
    class="input-body"
    :class="disabled ? 'disabled' : null"
    @click="focus()"
    :style="{
      backgroundColor: backgroundColor,
      borderRadius: radius,
      padding: padding,
    }"
  >
    <i :class="icon" :style="{ color: iconColor }"></i>
    <div style="position:relative;flex-grow:1;padding-right:1.2em;overflow:hidden;text-overflow:ellipsis">
      <div style="position:absolute;left:16px;text-overflow:ellipsis">
      <span
        style="letter-spacing: -0.02px;white-space:nowrap"
        :style="{
          fontSize: placeholderSize,
          fontWeight: placeholderWeight,
          top: `calc((80% - ${placeholderSize}) / 2)`,
          color: placeholderColor,
        }"
        v-show="input.length == 0"
        >{{ placeholder }}</span
      ></div>
      <input
        :disabled="disabled"
        :id="id"
        :style="{ color: color }"
        v-model="input"
        type="text"
      />
    </div>
  </div>
</template>
<script>
export default {
  name: "m-input",
  props: {
    icon: {
      type: String,
      default: "fas fa-search",
    },
    iconColor: {
      type: String,
      default: "rgba(0,0,0,0.62)",
    },
    placeholder: {
      type: String,
      default: "Enter text here",
    },
    placeholderColor: {
      type: String,
      default: "rgba(0,0,0,0.8)",
    },
    placeholderSize: {
      type: String,
      default: "14px",
    },
    placeholderWeight: {
      type: String,
      default: "normal",
    },
    color: {
      type: String,
      default: "rgba(0,0,0,.87)",
    },
    backgroundColor: {
      type: String,
      default: "#FFFFFF",
    },
    radius: {
      type: String,
      default: "5px",
    },
    padding: {
      type: String,
      default: "10px",
    },
    id: String,
    disabled: Boolean,
  },
  data() {
    return {
      input: "",
      showPlaceholder: true
    };
  },
  methods: {
    focus() {
      document.querySelector("#" + this.id).focus();
    },
  },
  watch: {
    input(val){
      this.$emit('input', val)
    }
  }
};
</script>
<style scoped>
input {
  border: none;
  background-color: rgba(0, 0, 0, 0);
  position: relative;
  padding-left: 16px;
  width: 100%;
  outline: none;
}
.input-body {
  display: flex;
  align-items: center;
  cursor: text;
  position:relative;
  overflow:hidden;
}
.input-body *{
  text-overflow:ellipsis !important;
}
.input-body.disabled {
  cursor: default;
}
</style>
