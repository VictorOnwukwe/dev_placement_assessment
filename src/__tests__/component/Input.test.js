import { mount } from "@vue/test-utils";
import Input from "../../components/Input";

describe("Text Box", () => {
  it("should emit its input value at key up", async () => {
    const wrapper = mount(Input);

    const input = wrapper.find("input");
    input.element.value = "good";
    await input.trigger("input");

    expect(wrapper.emitted().input[0]).toEqual(["good"]);
  });

  it("should render correctly when passed user prop", () => {
    const wrapper = mount(Input);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render placeholder text", () => {
      const wrapper = mount(Input, {propsData: {placeholder: "testing placeholder"}});

      expect(wrapper.text().includes("testing placeholder")).toBe(true);
  })

  describe("input", () => {
      it("should be disabled when disabled prop is true", () => {
        const wrapper = mount(Input, {propsData: {disabled: true}});

        expect(wrapper.find("input").attributes("disabled")).toBe("disabled");
      })
      it("should be enabled when disabled prop is false", () => {
        const wrapper = mount(Input, {propsData: {disabled: false}});

        expect(wrapper.find("input").attributes("disabled")).toBe(undefined);
      })
  })
});
