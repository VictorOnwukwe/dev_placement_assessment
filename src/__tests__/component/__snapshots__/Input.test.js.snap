// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Box should render correctly when passed user prop 1`] = `
<div class="input-body" style="background-color: rgb(255, 255, 255); border-radius: 5px; padding: 10px;"><i class="fas fa-search" style="color: rgba(0, 0, 0, 0.62);"></i>
  <div style="position: relative; flex-grow: 1; padding-right: 1.2em; overflow: hidden; text-overflow: ellipsis;">
    <div style="position: absolute; left: 16px; text-overflow: ellipsis;"><span style="letter-spacing: -0.02px; white-space: nowrap; font-size: 14px; font-weight: normal; color: rgba(0, 0, 0, 0.8);">Enter text here</span></div> <input type="text" style="color: rgba(0, 0, 0, 0.87);">
  </div>
</div>
`;
